tests/unit/js/test.subjects.js

```diff
@@ -0,0 +1,80 @@
+const sinon = require('sinon');
+const jquery = require('jquery');
+const { urlencode, Subject,
+    renderTag, slice } = require('../../../openlibrary/plugins/openlibrary/js/subjects');
+let sandbox, spy;
+
+beforeEach(() => {
+    sandbox = sinon.createSandbox();
+    global.$ = jquery;
+    spy = sinon.spy();
+    // Log is added inside utils.js, but we stub it here.
+    $.log = spy;
+    sandbox.stub(global, '$').callsFake(jquery);
+});
+
+describe('subjects: urlencode', () => {
+    test('empty array', () => {
+        expect(urlencode([])).toEqual('');
+    });
+    test('array of 1', () => {
+        expect(urlencode(['apple'])).toEqual('0=apple');
+    });
+    test('array of 3', () => {
+        expect(urlencode(['apple', 'grapes', 'orange'])).toEqual('0=apple&1=grapes&2=orange');
+    });
+});
+
+describe('subjects: urlencode', () => {
+    test('empty tag', () => {
+        expect(renderTag('', {id: "hello"}, '')).toEqual('< id="hello" ></>');
+    });
+    test('center tag', () => {
+        expect(renderTag('center', {id: "hello"}, '')).toEqual('<center id="hello" ></center>');
+    });
+    test('img tag', () => {
+        expect(renderTag('img', {id: "hello"}, '')).toEqual('<img id="hello" />');
+    });
+    test('img tag without no key values', () => {
+        expect(renderTag('img', {}, '')).toEqual('<img />');
+    });
+    test('img tag with multiple key values', () => {
+        expect(renderTag('img', {id: "hello", class: "bye"}, '')).toEqual('<img id="hello" class="bye" />');
+    });
+    test('img tag with key values and one child', () => {
+        expect(renderTag('img', {id: "hello", class: "bye"}, 'child')).toEqual('<img id="hello" class="bye" />');
+    });
+});
+
+describe('subjects: slice', () => {
+    test('empty array', () => {
+        expect(slice([], 0, 0)).toEqual([]);
+    });
+    test('array of 2', () => {
+        expect(slice([1, 2], 0, 1)).toEqual([1]);
+    });
+    test('arr length less than end', () => {
+        expect(slice([1, 2, 3], 0, 5)).toEqual([1, 2, 3]);
+    });
+    test('beginning greater than end', () => {
+        expect(slice([1, 2, 3, 4, 5], 4, 3)).toEqual([]);
+    });
+    test('array of 5', () => {
+        expect(slice([1, 2, 3, 4, 5], 0, 3)).toEqual([1, 2, 3]);
+    });
+});
+
+
+describe('subjects: Subject', () => {
+    test('constructor (not readable)', () => {
+        let data = { name: "Harry Potter/s+/g", works: [] };
+        const subject = new Subject(data, {});
+        expect(spy.calledOnce).toEqual(true);
```

Comment:
Took me a while to figure out what `spy` was; this is more readable.

```suggestion
        expect($.log.calledOnce).toEqual(true);
```

Reply:
$.log is just a reference to spy. I think it's better and more consistent with how sinon is used here to be explicit here to use spy.

Reply:
Then maybe let's call it `logSpy`? It's not uncommon for a test to have multiple spied on methods, so it's a little difficult to tell what this test is checking. Right now it just looks like "When the constructor is called, ensure that some method is called once", and I have to look up the reference to figure out what is being spied on.

Actually, now that I think about it, this is kind of a brittle test because it's checking that a log message is logged (which is weird). That's too specific to the implementation (unit tests should test contracts not implementations). The code will still work even if we remove that log statement, so it shouldn't cause a test to fail. Maybe let's just remove this line entirely?

Reply:
I think it's useful to record in the test that this feature triggers logging. I've renamed the variable.

Reply:
Closing the loop since we discussed this on slack:
It looked like `$.log` was an analytics method (which is why Jon was saying it makes sense to be tested), but the definition revealed to be a conditional `console.log` wrapper that was used twice, so Jon just removed the fn entirely :+1: